### ISSUE TITLE
haskell: cabal unable to find alex and happy in a shell produced with `shellFor`

### ISSUE THREAD

**Issue by d-xo** [NixOS/nixpkgs#130556]
**Describe the bug**

I have `alex` and `happy` listed in the `build-tool-depends` section of my cabal file ([here](https://github.com/ethereum/act/blob/master/src/act.cabal#L39)), but when I enter a nix shell created with `haskellPackages.shellFor` cabal is unable to find these dependencies and gives the following error:

```
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.RemoteRepo {remoteRepoName = RepoName
"hackage.haskell.org", remoteRepoURI = http://hackage.haskell.org/,
remoteRepoSecure = Just True, remoteRepoRootKeys =
["fe331502606802feac15e514d9b9ea83fee8b6ffef71335479a2e68d84adc6b0","1ea9ba32c526d1cc91ab5e5bd364ec5e9e8cb67179a471872f6e26f0ae773d42","2c6c3627bd6c982990239487f1abd02e08a02e6cf16edb105a8012d444d870c3","0a5c7ea47cd1b15f01f5f51a33adda7e655bc0f0b0615baa8e271f4c3351e21d","51f0161b906011b52c6613376b1ae937670da69322113a246a09f807c62f6921"],
remoteRepoKeyThreshold = 3, remoteRepoShouldTryHttps = True}
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: act-0.1.0.0 (user goal)
[__1] unknown package: act:happy:exe.happy (dependency of act)
[__1] fail (backjumping, conflict set: act, act:happy:exe.happy)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: act, act:happy:exe.happy
```

Both `alex` and `happy` are available on path in the shell:

```
> which alex
/nix/store/kw7250df32diiis9wx0x78irinnzibim-alex-3.2.6/bin/alex
> which happy
/nix/store/3k6bs1nz6bpkj9pfsfpmyqpfy0kjw5bd-happy-1.20.0/bin/happy
```

I can work around this by running `cabal v2-update` and then cabal will download and build `alex` and `happy` by itself, but I would prefer to avoid any impurities here and have cabal pull all it's dependencies from the nix-shell.

note that `haskell.nix` suffers from the same issue: https://github.com/input-output-hk/nix-tools/issues/95

cc @Ericson2314 @peti @ElvishJerricco

**Comment by matthewbauer** [comment]
I think this is from switching to new-style cabal builds. A temporary workaround is to always use `cabal v1-` in commands.

**Comment by d-xo** [comment]
Interesting! This does work :sparkles: :pray: I'm curious why that would be the case though?

**Comment by matthewbauer** [comment]
I think v2 is just stricter in a lot of places. Cabal wants the executable in the package database, but right now it doesn't get there.

**Comment by d-xo** [comment]
I see, do you know what would be required to get the executables into the package database in the nix shell? I would be interested to maybe try and make a PR here, but I'm very unfamiliar with the nix haskell infra and how cabal works in general...

**Comment by Ericson2314** [comment]
Yeah there some some issue about this before. There isn't a good way to get it in the package database. And for thing like is just an exe it doesn't make much since. It would be nicer if there was some other way to tell cabal "trust me, I'm the right version".

**Comment by d-xo** [comment]
Why shouldn't an executable go into the package db?

**Comment by sternenseemann** [comment]
Using `v1-*` commands is still the only workaround known to me.

**Comment by ParetoOptimalDev** [comment]
> Why shouldn't an executable go into the package db? It seems as though the v2 cabal workflows expect them to be there?

See this related issue: https://github.com/haskell/cabal/issues/8434

**Comment by ScottFreeCode** [comment]
I more complete but still minimal reproduction is here: https://gist.github.com/ScottFreeCode/87bec0c569598014053579990a5bef6e

I've collected GitHub issues relating to this problem here: https://gist.github.com/ScottFreeCode/ef9f254e2dd91544bba4a068852fc81f

**Comment by ScottFreeCode** [comment]
For those of you who got emailed immediately on the previous comment, the first gist with the reproduction should have been https://gist.github.com/ScottFreeCode/c1adb5681c5b61c61373615f63a4a7d2 – Sorry about that!

**Comment by deemp** [comment]
Another workaround is to not specify `alex` and `happy` in the `build-tools` and `build-tool-depends` and just provide them via Nix.

Inspired by https://github.com/haskell/cabal/issues/8434#issuecomment-1652017630

**Comment by deemp** [comment]
A more flexible workaround is to:

1. Put `build-tool-depends` and `build-tools` under a flag in the `.cabal` file.
1. Wrap `cabal-install` to set the flag in a `.nix` file.

In the `.cabal`:

```cabal
Flag nix
    Description: Nix provides `build-depends` and `build-depend-tools`
    Default: False
    Manual: True

...

library libName
    ...
    if !flag(nix)
      build-tool-depends:  alex:alex, happy:happy
```

In the `.nix`:

```nix
devTools =
    let
      output = config.haskellProjects.default.outputs;
      wrapTool =
        pkgsName: pname: flags:
        let
          pkg = pkgs.${pkgsName};
        in
        pkgs.symlinkJoin {
          name = pname;
          paths = [ pkg ];
          meta = pkg.meta;
          version = pkg.version;
          buildInputs = [ pkgs.makeWrapper ];
          postBuild = ''
            wrapProgram $out/bin/${pname} \
              --add-flags "${flags}"
          '';
        };
    in
    {
      inherit (output.finalPackages) alex happy;
      
      cabal = wrapTool "cabal-install" "cabal" "-v0 -fnix";

      ghc = builtins.head (
        builtins.filter (
          x: pkgs.lib.attrsets.isDerivation x && pkgs.lib.strings.hasPrefix "ghc-" x.name
        ) output.devShell.nativeBuildInputs
      );

      inherit (haskellPackages) haskell-language-server;
    };
```

Then, you can add all these devTools to your devShell.

If you want to make `cabal` see `happy` and `alex` outside of a devshell, you can add the `--prefix` argument:

```
wrapProgram $out/bin/${pname} \
     --add-flags "${flags}" \
     --prefix PATH : ${pkgs.lib.makeBinPath (with devTools; [ alex happy ])}
```

This code is a bit [haskell-flake](https://github.com/srid/haskell-flake)-specific.

You can replace `output.devShell` with a [shellFor](https://nixos.wiki/wiki/Haskell#Using_shellFor_.28multiple_packages.29) derivation.
